### PR TITLE
Handled retrial failure until max delay hitten due to OffsetOutOfRangeError by setting _fetch_offset to OFFSET_EARLIEST.

### PR DESCRIPTION
**Pull request by zsuzhengdu**
Inspired by https://github.com/dpkp/kafka-python/blob/1937ce59b4706b44091bb536a9b810ae657c3225/kafka/context.py#L169-L178

An offset fetch request will be triggered after setting the `_fetch_offset` to `EARLIEST`
```
 if (self._fetch_offset == OFFSET_EARLIEST or
                self._fetch_offset == OFFSET_LATEST):
            # We need to fetch the offset for our topic/partition
            offset_request = OffsetRequest(
                self.topic, self.partition, self._fetch_offset, 1)
            self._request_d = self.client.send_offset_request([offset_request])
            self._request_d.addCallbacks(
                self._handle_offset_response, self._handle_offset_error)
```

### CODE REVIEW

File: afkak/consumer.py [hunk 1]
```diff
@@ -733,6 +733,9 @@ def _handle_fetch_error(self, failure):
             # We've retried until we hit the max delay, log at warn
             log.warning("%r: Still failing fetching messages from kafka: %r",
                         self, failure)
+            if failure.check(OffsetOutOfRangeError):
+                log.info("Reset offset to earliest due to OffsetOutOfRangeError")
+                self._fetch_offset = OFFSET_EARLIEST
```
Comment (twm) [inline afkak/consumer.py]: I don't think that this is correct. Do you really only want to check for this error condition once we have retried enough to hit the max delay? And only every other attempt?

File: afkak/test/test_consumer.py [hunk 2]
```diff
@@ -550,14 +550,13 @@ def make_fail(*_, **__):
...
-                             [dbg_call] * 65 + [err_call])
-            self.assertEqual(klog.warning.mock_calls, [warn_call] * 34)
+                             [dbg_call] * (len(klog.debug.mock_calls) - 1) + [err_call])
         fetch_fail = mockback.mock_calls[0][1][0]
         assert isinstance(fetch_fail, Failure)
         fetch_fail.trap(KafkaUnavailableError)
         offset_call = call([request])
         self.assertEqual(mockclient.send_offset_request.mock_calls,
-                         [offset_call] * 100)
+                         [offset_call] * fetch_attempts)
```
Comment (twm) [inline afkak/test/test_consumer.py]: :+1:

**Comment by rthille** [comment]
Heh, I had notes (and I think a brief conversation with @twm) about how to implement a handler for this.  That is, I was thinking that the Consumer should be created with a specification for how to handle fetches where the server didn't have the messages the consumer was expecting.  I think there were about 6 cases to cover: Consumer has no stored offsets vs has stored offset; starting offset is specified explicitly vs stored vs earliest/latest; server has message offsets where the smallest is greater than what the consumer wants vs server has max offset < what the consumer wants.

**Comment by zsuzhengdu** [comment]
how to check the offset stored by consumer and where the consumer stored the offset?

**Comment by rthille** [comment]
IIRC, when you start() the consumer, you give it an offset specification (EARLIEST/LATEST/OFFSET_COMMITTED /<numeric-offset>).  So that's the consumer's notion of what offset it should next fetch.  In the case of OFFSET_COMMITTED, it has to do a pre-fetch fetch of the offset.  But once it has a numeric offset (either started with, or retrieved from the stored offsets), it can do a fetch of messages starting at that offset.  If that fails with OffsetOutOfRangeError, then it can look at the metadata associated with that topic to see what the range of offsets available is, and decide what to do based on configuration of the consumer (which doesn't exist yet).

**Comment by zsuzhengdu** [comment]
The failed test in Travis passed locally
```~/sources/afkak $ make toxc
KAFKA_VERSION=0.9.0.1 /Users/dzheng/sources/afkak/.env/bin/tox -i https://pypi.python.org/simple/ -e cur
GLOB sdist-make: /Users/dzheng/sources/afkak/setup.py
cur inst-nodeps: /Users/dzheng/sources/afkak/.tox/dist/afkak-3.0.0.dev0.zip
cur installed: afkak==3.0.0.dev0,attrs==18.1.0,Automat==0.7.0,constantly==15.1.0,coverage==4.0.1,funcsigs==1.0.2,hyperlink==18.0.0,idna==2.7,incremental==17.5.0,mock==1.3.0,nose==1.3.7,nose-timer==0.5.0,pbr==4.2.0,six==1.11.0,termcolor==1.1.0,Twisted==17.9.0,zope.interface==4.5.0
cur runtests: PYTHONHASHSEED='32899226'
cur runtests: commands[0] | nosetests --with-id --with-timer --timer-top-n 10 --with-coverage --cover-erase --cover-tests --cover-package afkak --logging-level=DEBUG --verbosity=2 --logging-format=%(asctime)s %(levelname)s %(filename)s %(lineno)d: %(message)s afkak.test.test_client_integration:TestAfkakClientIntegration.test_roundtrip_large_request
#250 afkak.test.test_client_integration.TestAfkakClientIntegration.test_roundtrip_large_request ... ok (0.5214s)

Name                                    Stmts   Miss  Cover   Missing
---------------------------------------------------------------------
afkak.py                                   13      0   100%
afkak/brokerclient.py                     171     36    79%   51, 129, 156, 161, 187, 191-192, 197-198, 223, 256-258, 276-294, 309, 320-324, 330-331, 346-347, 360, 368-369, 383, 415-418
afkak/client.py                           381    113    70%   81, 145-151, 163-165, 174, 177, 182, 210, 250-251, 279-281, 291-295, 311-357, 401, 427, 446-451, 460-466, 491-498, 509-521, 524, 564, 579-582, 590, 603-605, 641, 657, 660, 671-674, 686-695, 699-701, 748-749, 764-770, 809, 829-835, 882-885, 887, 901-902, 916, 925, 938-944
afkak/codec.py                             61     46    25%   7-8, 39, 45, 49, 53-57, 61-66, 94-114, 118-141
afkak/common.py                           133      4    97%   269, 295, 300, 302
afkak/consumer.py                         379    332    12%   131-203, 206, 236-257, 270-324, 335-381, 411-441, 446-447, 452-472, 476-479, 492-502, 514-533, 544-568, 571-572, 575-576, 579-580, 584-586, 592-595, 601-628, 640-671, 675-676, 690-692, 713-743, 756-836, 847-886, 901-948, 957-960, 965-972
afkak/kafkacodec.py                       261    102    61%   117, 140-154, 168, 178-186, 325-342, 351-366, 400, 451-457, 466-471, 488-508, 517-526, 539-554, 564-576, 599-604, 616-621, 641-652
afkak/partitioner.py                       87     66    24%   17-24, 51-106, 122-123, 134, 147, 150-151, 154, 158-162, 166-168, 177-183
afkak/producer.py                         254    217    15%   95-152, 155, 175-189, 195-205, 211-214, 223-225, 232-237, 247-272, 282-336, 345-352, 359-365, 382-408, 421-440, 464-608, 612-613, 617-619
afkak/protocol.py                          20      3    85%   38, 43-46
afkak/test.py                              12      1    92%   26
afkak/test/fixtures.py                    146     13    91%   123, 236-240, 248-254
afkak/test/service.py                      62      5    92%   81-88
afkak/test/test_client_integration.py     120     47    61%   82-91, 97-106, 116-126, 164-169, 181-220
afkak/test/testutil.py                    105     18    83%   39, 47, 64-67, 73, 105, 121, 165, 185-193, 197-199, 202-205
afkak/util.py                              47      8    83%   20, 23, 30, 34, 38, 46, 56, 65
---------------------------------------------------------------------
TOTAL                                    2252   1011    55%
afkak.test.test_client_integration.TestAfkakClientIntegration.test_roundtrip_large_request: 0.5214s
----------------------------------------------------------------------
Ran 1 test in 31.363s

OK```

File: afkak/consumer.py [hunk 3]
```diff
@@ -710,6 +710,13 @@ def _handle_fetch_error(self, failure):
         TODO: Possibly make this differentiate based on the failure
         """
         # The _request_d deferred has fired, clear it.
+        if failure.check(OffsetOutOfRangeError):
+            offset_request = OffsetRequest(
+                self.topic, self.partition, OFFSET_EARLIEST, 1)
```
Comment (rthille) [inline afkak/consumer.py]: This makes a big assumption. This should be configurable, since in some applications losing messages (because your afkak consumer was down until the messages expired) would be a critical error.  Here you're just going merrily along, assuming that the loss of any messages between what the consumer wanted (presumably a committed offset) and what's available can be ignored.

Comment (zsuzhengdu) [reply afkak/consumer.py]: I do make that assumption that messages between the consumer stored offset and minimal available off set of given topic is ignorable. 

What are the parameters to make `OffsetOutOfRange` error handling configurable? `handle_ OffsetOutOfRange` (True/False), `time` (-2/-1) and `max_offsets`?

Comment (zsuzhengdu) [reply afkak/consumer.py]: @rthille  Do you think provide `auto.offset.reset` a feasible solution which provides certain level user involvement or user control.

Comment (rthille) [reply afkak/consumer.py]: The Kafka-python approach seems reasonable, though I'd define 'earliest', 'latest', and 'raise' (explicitly), but that's just me.

Comment (zsuzhengdu) [reply afkak/consumer.py]: @twm  May I have your opinion as well?

File: afkak/test/test_consumer.py [hunk 4]
```diff
@@ -782,6 +782,20 @@ def test_consumer_error_during_offset(self):
...
+        topic = 'offset_out_of_range_error'
+        part = 911
+        reqs_ds = [Deferred(), Deferred()]
+        mockclient = Mock()
+        mockclient.send_offset_request.side_effect = reqs_ds
+        consumer = Consumer(mockclient, topic, part, Mock())
+        consumer._clock = MemoryReactorClock()
+        consumer.start(OFFSET_LATEST)
+        # Make sure request for offset was made
+        request = OffsetRequest(topic='error_during_offset', partition=part, time=-1, max_offsets=1)
```
Comment (rthille) [inline afkak/test/test_consumer.py]: The topic here looks wrong.  Also, it doesn't seem like the test does anything to test the offset being out of range?

File: afkak/test/test_consumer.py [hunk 5]
```diff
@@ -784,6 +782,27 @@ def test_consumer_error_during_offset(self):
...
+        d = consumer.start(offset)
+
+        fetch_request = FetchRequest(topic=topic, partition=part, offset=offset,
+                                     max_bytes=FETCH_BUFFER_SIZE_BYTES)
+        consumer.client.send_fetch_request.assert_called_once_with([fetch_request], max_wait_time=100, min_bytes=65536)
+
+        f = Failure(OffsetOutOfRangeError())
+        reqs_ds[0].errback(f)
+
+        # fetch offset
```
Comment (zsuzhengdu) [inline afkak/test/test_consumer.py]: @rthille  I need some help regarding mocking the trigger of OffsetOutOfRangeError()  to check the fire of or calling on `send_offset_request` with `fetch_request`.

File: afkak/test/test_consumer.py [hunk 6]
```diff
@@ -785,16 +785,23 @@ def test_consumer_error_during_offset(self):
     def test_consumer_offset_out_of_range_error(self):
         topic = 'offset_out_of_range_error'
         part = 911
+        offset = 10000
         reqs_ds = [Deferred(), Deferred()]
         mockclient = Mock()
         mockclient.send_offset_request.side_effect = reqs_ds
```
Comment (rthille) [inline afkak/test/test_consumer.py]: You're mocking the `send_offset_request` method on the client, but since now you're starting the consumer with a specific offset, `send_offset_request` shouldn't be called (at least not at first). Instead, you need to set the `send_fetch_request` side_effect to those deferreds so you can errback(f) them like you're doing below.  Then depending on how deep you want to do the mocking, you can either just check that `client.send_offset_request` was called with the proper request, or you can setup the side effects for that mock as well and make sure the consumer makes the proper fetch request with the proper offset.  I'd probably just check that the right `send_offset_request` was made after you `errback()` the send_fetch_request.

Comment (zsuzhengdu) [reply afkak/test/test_consumer.py]: Thank you so much @rthille  for the hints, which is appreciated!

**Comment by zsuzhengdu** [comment]
@rthille  I reran the same failed itest: `test_produce_request`; but can not reproduce.
```#6 afkak.test.test_client_integration.TestAfkakClientIntegration.test_produce_request ... ok (0.1526s)
```
May I know how to reproduce following exception on purpose (removed created znode...?)  or any work around I could do to make the failed tests pass which seems unrelated the changes made on consumer?
```2018-08-16 13:19:41,790 DEBUG service.py:73: [2018-08-16 13:19:41,351] INFO Got user-level KeeperException when processing sessionid:0x16542e40fbb0001 type:create cxid:0x215 zxid:0xdb txntype:-1 reqpath:n/a Error Path:/HmKoWHqDmG/brokers/topics/test_produce_large_request-ZKUuQrTvJE/partitions/1 Error:KeeperErrorCode = NoNode for /HmKoWHqDmG/brokers/topics/test_produce_large_request-ZKUuQrTvJE/partitions/1 (org.apache.zookeeper.server.PrepRequestProcessor)
2018-08-16 13:19:41,790 DEBUG service.py:73: [2018-08-16 13:19:41,790] INFO Got user-level KeeperException when processing sessionid:0x16542e40fbb0001 type:setData cxid:0x21c zxid:0xde txntype:-1 reqpath:n/a Error Path:/HmKoWHqDmG/config/topics/test_produce_request-uOjfqWzzLj Error:KeeperErrorCode = NoNode for /HmKoWHqDmG/config/topics/test_produce_request-uOjfqWzzLj (org.apache.zookeeper.server.PrepRequestProcessor)
2018-08-16 13:19:41,793 DEBUG service.py:73: [2018-08-16 13:19:41,790] INFO Got user-level KeeperException when processing sessionid:0x16542e40fbb0001 type:setData cxid:0x21c zxid:0xde txntype:-1 reqpath:n/a Error Path:/HmKoWHqDmG/config/topics/test_produce_request-uOjfqWzzLj Error:KeeperErrorCode = NoNode for /HmKoWHqDmG/config/topics/test_produce_request-uOjfqWzzLj (org.apache.zookeeper.server.PrepRequestProcessor)
2018-08-16 13:19:41,793 DEBUG service.py:73: [2018-08-16 13:19:41,793] INFO Got user-level KeeperException when processing sessionid:0x16542e40fbb0001 type:create cxid:0x21d zxid:0xdf txntype:-1 reqpath:n/a Error Path:/HmKoWHqDmG/config/topics Error:KeeperErrorCode = NodeExists for /HmKoWHqDmG/config/topics (org.apache.zookeeper.server.PrepRequestProcessor)```

File: afkak/consumer.py (hunk 3 above)
Comment (rthille) [reply afkak/consumer.py]: @twm ^^^

**Comment by rthille** [comment]
The test failures all look to be either "Kafka is 💩" (the 'not leader for partition' failures) or Jenkins running the tests on under-powered hardware.

File: afkak/consumer.py (hunk 3 above)
Comment (twm) [reply afkak/consumer.py]: Yes, I'd go for three options:

* EARLIEST
* LATEST
* FAIL ("fail" rather than "raise" as the failure will be reported on the Deferred returned by `start()`)

I also want to change the error handling within `Consumer` to only retry on a `BrokerResponseError` when it is retriable &mdash; see #23, which adds this metadata to the exceptions.

**Comment by rthille** [comment]
Just kicked the build again to see if they'll pass.

File: afkak/consumer.py [hunk 7]
```diff
@@ -180,7 +185,10 @@ def __init__(self, client, topic, partition, processor,
             raise ValueError(
                 'request_retry_max_attempts must be non-negative integer')
         self._fetch_attempt_count = 1
-
+        if not isinstance(auto_offset_reset, str) or auto_offset_reset not in ['earliest', 'latest', 'fail']:
+            raise ValueError(
+                'auto_offset_reset must be in \'earliest\', \'latest\', \'fail\'')
```
Comment (rthille) [inline afkak/consumer.py]: Switching quotes is usually cleaner:
`"auto_offset_reset must be in 'earliest', 'latest', 'fail'"`

File: afkak/consumer.py [hunk 8]
```diff
@@ -727,8 +735,19 @@ def _handle_fetch_error(self, failure):
             log.debug(
                 "%r: Exhausted attempts: %d fetching messages from kafka: %r",
                 self, self.request_retry_max_attempts, failure)
-            self._start_d.errback(failure)
-            return
+            if failure.check(OffsetOutOfRangeError):
```
Comment (rthille) [inline afkak/consumer.py]: You need to move this outside the `if (self.request_retry_max_attempts != 0 and` check at line 725/733, or in the default case it'll never trigger.

File: afkak/consumer.py [hunk 9]
```diff
@@ -128,7 +132,8 @@ def __init__(self, client, topic, partition, processor,
                  max_buffer_size=None,
                  request_retry_init_delay=REQUEST_RETRY_MIN_DELAY,
                  request_retry_max_delay=REQUEST_RETRY_MAX_DELAY,
-                 request_retry_max_attempts=0):
+                 request_retry_max_attempts=0,
+                 auto_offset_reset='earliest'):
```
Comment (rthille) [inline afkak/consumer.py]: This should default to 'fail', since having the wrong offset as a consumer is a serious, data-losing error.

File: afkak/consumer.py [hunk 10]
```diff
@@ -718,14 +718,21 @@ def _handle_fetch_error(self, failure):
...
+            elif self.auto_offset_reset == 'latest':
+                offset_request = OffsetRequest(
+                    self.topic, self.partition, OFFSET_LATEST, 1)
+            else:
+                self._start_d.errback(failure)
+                return
             self._request_d = self.client.send_offset_request([offset_request])
-            self._request_d.addCallbacks(
-                self._handle_offset_response, self._handle_offset_error)
+            self._request_d.addCallbacks(self._handle_offset_response, self._handle_offset_error)
```
Comment (rthille) [inline afkak/consumer.py]: I see that I've used this form in a least a couple of places, but I think that this form:
https://github.com/ciena/afkak/blob/master/afkak/consumer.py#L937-L941
may be better, because if there's a coding error in _handle_offset_response, then the _handle_offset_error code can catch it.  With the addCallbacks form, I don't the error-handler "side" can handle errors in the success "side".
@twm Probably has a more well-formed opinion on that.  But note that there's no down-stream error handling on the consumer's private `self._request.d`.

**Comment by zsuzhengdu** [comment]
@rthille There is a test case named `test_consumer_fetch_retry_to_failure` which never finished running as shown in https://travis-ci.org/ciena/afkak/jobs/443729826.

I could reproduce it locally. Can please provide me some clue like how to debug it or fix it?

File: afkak/consumer.py [hunk 11]
```diff
@@ -720,8 +744,7 @@ def _handle_fetch_error(self, failure):
             log.debug(
                 "%r: Exhausted attempts: %d fetching messages from kafka: %r",
                 self, self.request_retry_max_attempts, failure)
-            self._start_d.errback(failure)
-            return
+
```
Comment (rthille) [inline afkak/consumer.py]: This change (above) is why that test is failing.  The consumer has decided that it's run out of attempts, but you removed the errback() call and the early return.

Comment (zsuzhengdu) [reply afkak/consumer.py]: Thanks @rthille!

File: afkak/consumer.py [hunk 12]
```diff
@@ -208,7 +208,7 @@ def __init__(self, client, topic, partition, processor,
         self._msg_block_d = None  # deferred for each block of messages
         self._processor_d = None  # deferred for a result from processor
         self._state = '[initialized]'  # Keep track of state for debugging
-        # Check parameters for sanity
+        # Check parameters for sanityt
```
Comment (rthille) [inline afkak/consumer.py]: Typo.

**Comment by zsuzhengdu** [comment]
Change `download-kafka.py` to `download-kafka` on your box? ?

Wried, either way works for me on Python 2.7

```py35-lint create: /Users/dzheng/sources/dzheng/afkak/.tox/py35-lint
ERROR: InterpreterNotFound: python3.5
py35-unit create: /Users/dzheng/sources/dzheng/afkak/.tox/py35-unit
ERROR: InterpreterNotFound: python3.5
py35-unit-snappy create: /Users/dzheng/sources/dzheng/afkak/.tox/py35-unit-snappy
ERROR: InterpreterNotFound: python3.5
py35-int-snappy create: /Users/dzheng/sources/dzheng/afkak/.tox/py35-int-snappy
ERROR: InterpreterNotFound: python3.5
__________________________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________________________
  py27-lint: commands succeeded
  py27-unit: commands succeeded
  py27-unit-snappy: commands succeeded
  py27-unit-murmur: commands succeeded
  py27-int-snappy-murmur: commands succeeded
  pypy-lint: commands succeeded
  pypy-unit: commands succeeded
  pypy-unit-snappy: commands succeeded
  pypy-unit-murmur: commands succeeded
  pypy-int-snappy-murmur: commands succeeded
ERROR:   py35-lint: InterpreterNotFound: python3.5
ERROR:   py35-unit: InterpreterNotFound: python3.5
ERROR:   py35-unit-snappy: InterpreterNotFound: python3.5
ERROR:   py35-int-snappy: InterpreterNotFound: python3.5
make: *** [toxa] Error 1```

**Comment by rthille** [comment]
Yeah, I fixed the makefile, and then it failed because tox didn't exist (because there was no dependency to create it, so I added that, and then the build failed to install python-snappy because I didn't have snappy installed, nor pypy, so I'm waiting for those to finish installing before I run into the next error :-)

**Comment by rthille** [comment]
Hey @twm any reason why you're sticking to Python 3.5, rather than 3.7?

**Comment by rthille** [comment]
```
py35-lint create: /Users/rthille/dev/python/afkak/.tox/py35-lint
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError("Failed to get version_info for python3.5: pyenv: python3.5: command not found\n\nThe `python3.5' command exists in these Python versions:\n  3.5.6\n\n",)
py35-unit create: /Users/rthille/dev/python/afkak/.tox/py35-unit
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError("Failed to get version_info for python3.5: pyenv: python3.5: command not found\n\nThe `python3.5' command exists in these Python versions:\n  3.5.6\n\n",)
py35-unit-snappy create: /Users/rthille/dev/python/afkak/.tox/py35-unit-snappy
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError("Failed to get version_info for python3.5: pyenv: python3.5: command not found\n\nThe `python3.5' command exists in these Python versions:\n  3.5.6\n\n",)
py35-int-snappy create: /Users/rthille/dev/python/afkak/.tox/py35-int-snappy
ERROR: Error creating virtualenv. Note that some special characters (e.g. ':' and unicode symbols) in paths are not supported by virtualenv. Error details: InvocationError("Failed to get version_info for python3.5: pyenv: python3.5: command not found\n\nThe `python3.5' command exists in these Python versions:\n  3.5.6\n\n",)
_________________________________________________________________ summary __________________________________________________________________
```
That's after using pyenv to install python 3.5.6. Ugh python is terrible.

**Comment by rthille** [comment]
Figured out a work around. Install 2 versions of python with pyenv and activate them both with:
```
pyenv install 2.7.15
pyenv install 3.5.6
pyenv local 2.7.15 3.5.6
```
Then with the shims dir in my path then 'python3.5' works.

**Comment by twm** [comment]
@rthille We still deploy to Python 3.5, so we still need to support it. It is also what's OS-available on my desktop.

Twisted and Travis didn't support 3.7 until quite recently, but I would be happy to accept a (separate) PR which adds it to the Tox matrix. https://github.com/twm/yarrharr/pull/276 may provide useful information.

**Comment by zsuzhengdu** [comment]
@rthille @twm Can we have this merged and have a new release please, which is needed for a ciena internal issue?

**Comment by rthille** [comment]
```
__________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________
  py27-lint: commands succeeded
  py27-unit: commands succeeded
  py27-unit-snappy: commands succeeded
  py27-unit-murmur: commands succeeded
  py27-int-snappy-murmur: commands succeeded
  pypy-lint: commands succeeded
  pypy-unit: commands succeeded
  pypy-unit-snappy: commands succeeded
  pypy-unit-murmur: commands succeeded
  pypy-int-snappy-murmur: commands succeeded
  py35-lint: commands succeeded
  py35-unit: commands succeeded
  py35-unit-snappy: commands succeeded
  py35-int-snappy: commands succeeded
  congratulations :)
Done
---( Make git-clean all Complete (time: 928s) )---
      927.22 real       594.47 user       111.24 sys
```
Woot!  Still no coverage report though, and 'make toxcov' seems borked :-)

**Comment by rthille** [comment]
> @rthille @twm Can we have this merged and have a new release please, which is needed for a ciena internal issue?

I'm happy with the changes. Not sure they're correct, and and @twm "owns" Afkak (at least as far as Ciena is concerned), so he's the guy you've got to convince :-)

**Comment by rthille** [comment]
So I got coverage to work with some small changes:
```
_________________________________________________________________ summary __________________________________________________________________
  py35-int-snappy-coverage: commands succeeded
  congratulations :)
GLOB sdist-make: /Users/rthille/dev/python/afkak/setup.py
cov_report installed: attrs==18.2.0,Automat==0.7.0,constantly==15.1.0,coverage==4.0.1,funcsigs==1.0.2,hyperlink==18.0.0,idna==2.7,incremental==17.5.0,mock==1.3.0,nose==1.3.7,nose-timer==0.5.0,pbr==5.0.0,PyHamcrest==1.9.0,six==1.11.0,termcolor==1.1.0,Twisted==18.7.0,zope.interface==4.5.0
cov_report runtests: PYTHONHASHSEED='1761195416'
cov_report runtests: commands[0] | coverage combine
cov_report runtests: commands[1] | coverage report
Name                                         Stmts   Miss Branch BrPart  Cover
------------------------------------------------------------------------------
afkak/__init__.py                               13      0      0      0   100%
afkak/_util.py                                  81      7     44      7    89%
afkak/brokerclient.py                          146      0     38      2    99%
afkak/client.py                                390      1    134      4    99%
afkak/codec.py                                  61      0     14      0   100%
afkak/common.py                                313      0      8      0   100%
afkak/consumer.py                              389      6    176      8    97%
afkak/kafkacodec.py                            267      2     86      2    99%
afkak/partitioner.py                            87      0     22      0   100%
afkak/producer.py                              259      0    105      2    99%
afkak/protocol.py                               20      0      2      0   100%
afkak/test/__init__.py                          19      5      4      2    70%
afkak/test/fixtures.py                         165      1      4      0    99%
afkak/test/service.py                           77     11     22      3    86%
afkak/test/test_brokerclient.py                404      0      4      0   100%
afkak/test/test_client.py                      828      0     73      2    99%
afkak/test/test_client_integration.py           99      0      0      0   100%
afkak/test/test_codec.py                        43      0      4      0   100%
afkak/test/test_common.py                       43      0     20      0   100%
afkak/test/test_consumer.py                   1056      0     22      1    99%
afkak/test/test_consumer_integration.py        138      0     18      0   100%
afkak/test/test_failover_integration.py         88      0     16      2    98%
afkak/test/test_kafkacodec.py                  274      0     10      0   100%
afkak/test/test_package.py                      22      0      0      0   100%
afkak/test/test_partitioner.py                 130      0     22      0   100%
afkak/test/test_performance_integration.py     107      0     12      1    99%
afkak/test/test_producer.py                    493      0     10      1    99%
afkak/test/test_producer_integration.py        322      0     18      2    99%
afkak/test/test_protocol.py                     44      0      0      0   100%
afkak/test/test_util.py                         49      0      0      0   100%
afkak/test/testutil.py                         113      2     30      3    97%
------------------------------------------------------------------------------
TOTAL                                         6540     35    918     42    99%
cov_report runtests: commands[2] | coverage html -d /Users/rthille/dev/python/afkak/htmlcov
cov_report runtests: commands[3] | coverage xml -o /Users/rthille/dev/python/afkak/coverage.xml
_________________________________________________________________ summary __________________________________________________________________
  cov_report: commands succeeded
  congratulations :)
make: *** [toxcov] Error 1
      795.79 real       549.42 user        79.22 sys

Compilation exited abnormally with code 2 at Tue Oct 23 11:52:41

```
Looks like coverage has fallen a bit. Maybe if we can get it back in the build and just make sure it doesn't get worse that'd be the best approach.

File: afkak/consumer.py [hunk 13]
```diff
@@ -180,7 +185,10 @@ def __init__(self, client, topic, partition, processor,
             raise ValueError(
                 'request_retry_max_attempts must be non-negative integer')
         self._fetch_attempt_count = 1
-
+        if not isinstance(auto_offset_reset, str) or auto_offset_reset not in ['earliest', 'latest', 'fail']:
```
Comment (twm) [inline afkak/consumer.py]: It's fine to omit the type check, then you don't need a test for it.
```suggestion
        if auto_offset_reset not in {'earliest', 'latest', 'fail'}:
```

**Comment by twm** [comment]
Hi @zsuzhengdu, I'm sorry it has taken me so long to get to this. I don't have time for a full review at the moment, but I did a skim and have a few notes:

1. This feature addition merits an entry in the changelog.
2. There are two new behaviors here &mdash; earliest and latest &mdash; but only one new test. We need all these new paths to be covered.
3. It seems like we could use the existing constants `OFFSET_EARLIEST` and `OFFSET_LATEST` for the values passed to `auto_offset_reset`. `'fail'` could be represented by ``None``.

Thanks for your patience!

**Comment by zsuzhengdu** [comment]
@twm ^^
